### PR TITLE
Add _Nullable and _Nonnull type qualifiers

### PR DESCRIPTION
**Pull request by acowley**

### CODE REVIEW

**Comment by visq** [comment]
thanks a lot for the patch!
did you also consider the type checker, i.e. make sure all pattern matches are exhaustive?
I'm on holiday until 25/6 with limited email access, so please be patient.
Benedikt

**Comment by acowley** [comment]
I'll do a `-Wall` build to see if I missed any. This change is needed for Darwin, as usual. There are `c2hs` failures with `OpenCL` which are addressed with this change. So I've tested that path through the code, but nothing else; I'll do the `-Wall` build later today.

**Comment by acowley** [comment]
I did a big -Wall audit; should I drop that commit in this PR?

I didn't address everything, as some of it is arguably stylistic choice (e.g. there is a fair bit of name shadowing in the lexer), but I still hit nearly every file. There are still some non-exhaustive pattern matches, but they are not to do with the `_Nullable` and `_Nonnull` qualifiers, so I did not address them all as I don't know what the right thing to do is in every case.

Assuming you're okay with this broad set of superficial changes, I think we should really setup some CI to make sure we don't break compatibility with older GHCs while doing cleanup. Can you hit the Travis switch for the repo?

**Comment by acowley** [comment]
I pushed the commit. I tried to leave out purely stylistic changes like removing redundant brackets, but I did make a few changes that seemed double-redundant. The `c2hs` test suite passes with this version and GHC-7.10.3, but it would be good to test things out in a variety of configurations.

**Comment by visq** [comment]
Hi Anthony, thanks a lot for your work and fixing warnings! I also started working on support for C11 a while ago, so now I'm in the process of merging the commits. Setting up automated tests is a great idea, I'm all for it but had not yet time to look into it.

File: src/Language/C/Analysis/TravMonad.hs [hunk 1]
```diff
@@ -356,7 +354,7 @@ handleTravError a = liftM Just a `catchTravError` (\e -> recordError e >> return
 
 -- | check wheter non-recoverable errors occurred
 hadHardErrors :: [CError] -> Bool
-hadHardErrors = (not . null . filter isHardError)
+hadHardErrors = not . any isHardError
```
Comment (visq) [inline src/Language/C/Analysis/TravMonad.hs]: That should be hadHardErrors = any isHardError ;)
Fixed during merging.

Comment (acowley) [inline src/Language/C/Analysis/TravMonad.hs]: Gah, good catch!

**Pull request merged**